### PR TITLE
Added Support for Gpsd

### PR DESCRIPTION
**Pull request by salvatore-coppola**
Added Support for Gpsd through `de.taimos:gpsd4java:1.10`

### CODE REVIEW

File: kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/position/PositionService.java
Comment (MMaiero) [inline kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/position/PositionService.java:108]: Do we have an API method to use instead? If so, maybe let's reference it here.

Comment (MMaiero) [reply kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/position/PositionService.java:108]: Le's add also from when the API has been deprecated. It will help also with removals

File: kura/org.eclipse.kura.linux.position/OSGI-INF/gpsdPositionPositionProvider.xml
Comment (MMaiero) [inline kura/org.eclipse.kura.linux.position/OSGI-INF/gpsdPositionPositionProvider.xml]: The name of the component definition file has position 2 times. Is that expected?

File: kura/org.eclipse.kura.linux.position/OSGI-INF/serialDevicePositionProvider.xml
Comment (MMaiero) [inline kura/org.eclipse.kura.linux.position/OSGI-INF/serialDevicePositionProvider.xml]: It's a new file so the copyright header should have only 2022

File: kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java
Comment (MMaiero) [inline kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java]: Only 2022 since it is a new file

File: kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/PositionProvider.java
Comment (MMaiero) [inline kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/PositionProvider.java:14]: copyright header missing

File: kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/PositionProvider.java
Comment (MMaiero) [inline kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/PositionProvider.java]: If this is a new interface, not clear the deprecations.

File: kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/PositionProviderType.java
Comment (MMaiero) [inline kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/PositionProviderType.java:14]: copyright header missing

File: kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/SerialDevicePositionProvider.java
Comment (MMaiero) [inline kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/SerialDevicePositionProvider.java]: 2022 if this is a new file

File: kura/test/org.eclipse.kura.linux.position.test/src/test/resources/gpsd-raw-json-boltgate-10-12-el27.1.txt
Comment (MMaiero) [inline kura/test/org.eclipse.kura.linux.position.test/src/test/resources/gpsd-raw-json-boltgate-10-12-el27.1.txt]: I would anonymise the name of such files

File: kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/position/PositionService.java
Comment (salvatore-coppola) [reply kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/position/PositionService.java:108]: > Do we have an API method to use instead? If so, maybe let's reference it here.

No, the idea is to now allow to access directly to the underlying position provider.

Comment (salvatore-coppola) [reply kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/position/PositionService.java:108]: > Le's add also from when the API has been deprecated. It will help also with removals

I've added since 2.2.0

File: kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/PositionProvider.java
Comment (salvatore-coppola) [reply kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/PositionProvider.java]: to maintain the  compatibility with SerialDevicePositionProvider  than now implements Position Provider.

Comment (nicolatimeus) [reply kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/PositionProvider.java]: Maybe we can split this in two internal interfaces? For example PositionProvider, which only allows to get the position and time, and NMEAPositionProvider that provides the NMEA-related informations.

File: kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java
Comment (nicolatimeus) [inline kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java]: Could it make sense to define an internal "State" class containing all of these fields? In this way we can keep a single `AtomicReference<State>` here and have the whole state updated atomically. Otherwise it is possible that different fields come from different `TPVObject` instances due to race conditions.

File: kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/PositionProvider.java
Comment (salvatore-coppola) [reply kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/PositionProvider.java]: > Maybe we can split this in two internal interfaces? For example PositionProvider, which only allows to get the position and time, and NMEAPositionProvider that provides the NMEA-related informations.

I can do that, but also GpsdPositionProvider has to implement NMEAPositionProvider interface because PositionServiceImpl invoke the getNMEAPosition, unless i do a check on runtime class that I find an ugly solution.

File: kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java
Comment (salvatore-coppola) [reply kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java]: Looking at the current Gpsd4Java implementation it seems hard the race condition will happen, however it is a good idea. In the future we have to notice if a field is related to the others before add it in the internal state.

File: kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java
Comment (nicolatimeus) [inline kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java]: Should we create a local variable such as
```
final GpsdInternalState currentState = this.internalState.get();
```
and then get the fields out of it? in this way we are sure that all measurements come from the same `TPVObject`.

File: kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java
Comment (nicolatimeus) [inline kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java]: Probably we should create a new `GpsdInternalState` instance here, fill it from the `TPVObject` and then "publish" it atomically with `internalState.set()`.

File: kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/position/PositionService.java
Comment (salvatore-coppola) [reply kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/position/PositionService.java:108]: deprecated annotation removed

File: kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java
Comment (salvatore-coppola) [reply kura/org.eclipse.kura.linux.position/src/main/java/org/eclipse/kura/linux/position/GpsdPositionProvider.java]: done!

**Comment by MMaiero** [comment]
Build is failing but not due to tests introduced by this PR

File: kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/position/PositionService.java
Comment (MMaiero) [inline kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/position/PositionService.java]: Is it deprecated since 2.2 or 2.3?

File: target-platform/p2-repo-common/pom.xml
Comment (nicolatimeus) [inline target-platform/p2-repo-common/pom.xml]: Is this correct?

**Review by MMaiero — APPROVED** [review]
The contribution is above 1000 lines. But contains also gps test values useful for feature testing and that have raised the size of the PR.

**Review by nicolatimeus — APPROVED** [review]

**Comment by MMaiero** [comment]
The CQ for the gpsd dependency is still open https://dev.eclipse.org/ipzilla/show_bug.cgi?id=23923 but we have the PMC approval and, due to the new Eclipse rules, we will merge the contribution, keeping it on hold for releases until full approval by the Eclipse IP Team

**Pull request merged**
